plugins/outputs/azuremonitor/azuremonitor.go

```diff
@@ -0,0 +1,445 @@
+	}
+
+	err = a.validateCredentials()
+	if err != nil {
+		return fmt.Errorf("E! Unable to fetch authentication credentials: %v", err)
+	}
+
+	a.Reset()
+
+	return nil
+}
+
+func (a *AzureMonitor) validateCredentials() error {
+	if a.useMsi {
+		// Check expiry on the token
+		if a.msiToken == nil || a.msiToken.expiresInDuration() < time.Minute {
+			msiToken, err := a.getMsiToken(a.AzureClientID)
+			if err != nil {
+				return err
+			}
+			a.msiToken = msiToken
+		}
+		return nil
+	}
+
+	adToken, err := adal.NewServicePrincipalToken(
+		*(a.oauthConfig), a.AzureClientID, a.AzureClientSecret,
+		azure.PublicCloud.ActiveDirectoryEndpoint)
+	if err != nil {
+		return fmt.Errorf("Could not acquire ADAL token: %s", err)
+	}
+	a.adalToken = adToken
+	return nil
+}
+
+// Close shuts down an any active connections
+func (a *AzureMonitor) Close() error {
+	a.client = nil
+	return nil
+}
+
+type azureMonitorMetric struct {
+	Time time.Time         `json:"time"`
+	Data *azureMonitorData `json:"data"`
+}
+
+type azureMonitorData struct {
+	BaseData *azureMonitorBaseData `json:"baseData"`
+}
+
+type azureMonitorBaseData struct {
+	Metric         string                `json:"metric"`
+	Namespace      string                `json:"namespace"`
+	DimensionNames []string              `json:"dimNames"`
+	Series         []*azureMonitorSeries `json:"series"`
+}
+
+type azureMonitorSeries struct {
+	DimensionValues []string `json:"dimValues"`
+	Min             float64  `json:"min"`
+	Max             float64  `json:"max"`
+	Sum             float64  `json:"sum"`
+	Count           int64    `json:"count"`
+}
+
+// Write writes metrics to the remote endpoint
+func (a *AzureMonitor) Write(metrics []telegraf.Metric) error {
+	azmetrics := make(map[uint64]*azureMonitorMetric, len(metrics))
+	for _, m := range metrics {
+		id := hashIDWithTagKeysOnly(m)
+		if azm, ok := azmetrics[id]; !ok {
+			azmetrics[id] = translate(m)
+		} else {
+			azmetrics[id].Data.BaseData.Series = append(
+				azm.Data.BaseData.Series,
+				translate(m).Data.BaseData.Series...,
+			)
+		}
+	}
+
+	var body []byte
+	for _, m := range azmetrics {
+		// Azure Monitor accepts new batches of points in new-line delimited
+		// JSON, following RFC 4288 (see https://github.com/ndjson/ndjson-spec).
+		jsonBytes, err := json.Marshal(&m)
+		if err != nil {
+			return err
+		}
+		body = append(body, jsonBytes...)
+		body = append(body, '\n')
+	}
+
+	if err := a.validateCredentials(); err != nil {
+		return fmt.Errorf("E! Unable to fetch authentication credentials: %v", err)
+	}
+
+	req, err := http.NewRequest("POST", a.url, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+a.msiToken.AccessToken)
+	req.Header.Set("Content-Type", "application/x-ndjson")
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 || resp.StatusCode < 200 {
```

Comment:
Why are we only doing fmt.Errorf for status codes in the [300, 200) range.  We should also follow this pattern for any errors encountered in the 4xx or 5xx range